pymeasure/instruments/keithley/keithley2281S.py

```diff
@@ -0,0 +1,537 @@
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+import enum
+import time
+
+import numpy as np
+import pandas as pd
+
+from pymeasure.instruments import Instrument, SCPIMixin
+from pymeasure.instruments.keithley.buffer import KeithleyBuffer
+from pymeasure.instruments.validators import strict_discrete_set, strict_range, truncated_range
+
+
+@enum.unique
+class Keithley2281SOperationEventRegister(enum.IntFlag):
+    """
+    Enum containing Keithley2281S Operation Instrument Summary Event Register definition
+    """
+
+    CALIBRATION = 1  # Performing calibration
+    _RESERVED_1 = 2
+    _RESERVED_2 = 4
+    _RESERVED_3 = 8
+    MEASUREMENT = 16  # Performing measurement
+    TRIGGER = 32  # Waiting for trigger
+    ARM = 64  # Waiting for arm event
+    _RESERVED_4 = 128
+    FILT = 256  # Filter has settled or is disabled
+    LIST = 512  # Running a list
+    IDLE = 1024  # Idling
+    _RESERVED_5 = 2048
+    _RESERVED_6 = 4096
+    _RESERVED_7 = 8192
+    _RESERVED_8 = 16384
+
+
+@enum.unique
+class Keithley2281SMeasurementEventRegister(enum.IntFlag):
+    """
+    Enum containing Keithley2281S Measurement Instrument Summary Event Register definition
+    """
+
+    ROF = 1  # Reading overflow
+    _RESERVED_1 = 2
+    _RESERVED_2 = 4
+    _RESERVED_3 = 8
+    _RESERVED_4 = 16
+    _RESERVED_5 = 32
+    RAV = 64  # Reading available
+    BHF = 128  # Buffer half-full
+    BH = 256  # Buffer full
+    BTF = 512  # Buffer three-quarters full
+    BQF = 1024  # Buffer one-quarter-full
+    _RESERVED_6 = 2048
+    _RESERVED_7 = 4096
+    _RESERVED_8 = 8192
+    _RESERVED_9 = 16384
+
+
+class Keithley2281S(SCPIMixin, Instrument, KeithleyBuffer):
+    """
+    Represents the Keithley 2281S-20-6 power supply and battery simulator / characterizer.
+    Common commands beside `function_mode` and power supply commands should also work for
+    Keithley 2280S power supplies, although this is untested.
+    """
+
+    _VOLTAGE_RANGE = [0.0, 20]
+    _CURRENT_RANGE_PS = [0.1, 6.1]
+    _CURRENT_RANGE_BT_BS = [0.0, 6.1]
+    _INTERNAL_MEMORY_SLOTS = [*range(1, 10)]
+    _PLC_RANGE = [0.002, 12]
+
+    def __init__(self, adapter, name="Keithley2281S", **kwargs):
+        super().__init__(adapter, name, **kwargs)
+        if self.cm_line_frequency == 60:
+            self._PLC_RANGE[1] = 15
+
+    # Overwritten parent property
+
+    @property
+    def buffer_data(self):
+        """Get a pandas dataframe of values from the buffer."""
+        # Replace with match case, once Python>=3.10 is required
+        function_mode = self.cm_function_mode
+        if function_mode == "POWER":
+            return self.ps_buffer_data
+        elif function_mode == "TEST":
+            return self.bt_buffer_data
+        elif function_mode == "SIMULATOR":
+            return self.bs_buffer_data
+
+    # Common commands (cm_*), these can be used in any function mode
```

Comment:
Did you consider using channels for the power supply, battery testing, and battery simulation modes?

That would make it easier to understand, I think, as you could drop all the prefixes.

Reply:
Good idea, i will look into channels. They must be mutually exclusive thou.

Reply:
You could add/remove the channels dynamically.
However, I think it is best to keep them active, as the driver never know, which mode the device is in.
Right now, you do not disable these "forbidden" methods either.

Reply:
By the way: I thought about three different channels: One simulation, one measurement and one power supply.

Reply:
Finally came around to implement the channels. Unfortunately I had to put some constants in the global scope of the file, since I could not call `self.parent.var` from the channel's properties for some reason.   
The `_PLC_RANGE` 'constant' might be changed once on startup, if a 60Hz line-frequency is detected. Maybe the global scope might be a problem in the edge case where one uses two of these devices in the same script and one would be on 50Hz and the other on 60Hz. But I guess this is a very exotic use-case.

Reply:
Don't worry about global constants. That's totally fine as the user will never see them.

Reply:
Allright. I will do some last tests on my device at work next week and see if it works after the addition of channels and then this can be merged I think.

Reply:
Sorry for the long delay. I could finally test all the functions with a device and fix or remove what did not work.
The code is now in a state where it can be shared with others. The latest master branch is merged and all the checks are run as well.